File: apps/algo/src/main/java/sort/InsertionSort.java
package sort;

public class InsertionSort implements BiDirectionalIntSorter {

  @Override
  public int[] sort(int[] input) {
    for (int currId = 1; currId < input.length; currId++) {
      int element = input[currId];
      int previousId = currId - 1;

      while (previousId >= 0 && input[previousId] > element) {
        input[previousId + 1] = input[previousId];
        previousId--;
      }
      input[previousId + 1] = element;
    }
    return input;
  }

  @Override
  public int[] sortDesc(int[] input) {
    for (int current = 1; current < input.length; current++) {
      int element = input[current];
      int previous = current - 1;

      while (previous >= 0 && input[previous] < element) {
        input[previous + 1] = input[previous];
        previous--;
      }
      input[previous + 1] = element;
    }
    return input;
  }

}

File: books/algo/sort/insertion_sort.go
package sort

func insertionSort(arr []int) {
	for k := 1; k < len(arr); k++ {
		key := arr[k]
		p := k - 1
		for p >= 0 && arr[p] > key {
			arr[p+1] = arr[p]
			p--
		}
		arr[p+1] = key
	}
}
